2015/intro_concurrency.slide

```diff
@@ -0,0 +1,375 @@
+
+* Why Go:
+Go is scalable:
+
+- [[http://talks.golang.org/2012/splash.article][Language Design in the Service of Software Engineering]]
+- Built for large programs and large teams
+- Composition (types, functions, packages)
+
+* Why Go
+Go is fast:
+
+- Goal within 10% of C
+- Short compile times
+- Easy parallel computing (more later)
+
+* Why Go:
+Go is easy (language design):
+
+- Simple syntax (discussed later)
+- Garbage collected (no automatic memory management)
+- Stack traces on nil dereference (not "segmentation fault")
+- Pointers but automatically on stack wherever possible
+- Composition of types, not inheritance
+- Consistent design - legible code
+- Large standard library
+
+* Why Go
+Go is easy (tools):
+
+- Dowloading packages (go get)
+- No makefiles (go install)
+- Testing suite (go test)
+- Race/Deadlock detection (go install -race)
+- Automatic documentation (godoc)
+- Code formatting (gofmt)
+- Automatic import detection (goimports)
+- Profiling (go tool pprof)
+
+* Why Go
+Better than any other language to (personal opinion):
+
+- Effortlessly use external libraries
+- Manage Complexity
+- Write legible performant code
+- Write code that will last
+
+Individual steps may require more work, but stays managable as the program grows
+
+* Numeric programming
+[[github.com/gonum][Gonum]] project building a core of numeric functions
+
+- BLAS implementation (both native Go and C bindings)
+- Good matrix library (and constantly getting better)
+- Stats, finite difference, graphs etc.
+- Unconstrained local optimization (constrained and global under development)
+
+Not as fully-featured as scipy, but a new consistent design all in pure Go
+
+* Whirlwind tour of go
+
+* Hello, world!
+.play intro_concurrency/helloworld.go
+
+Run with:
+
+  go run helloworld.go
+
+Compile:
+  
+  go build helloworld.go
+
+Compile and place on $PATH:
+
+  go install helloworld.go
+
+* Variable and Function Declarations
+.play intro_concurrency/basics.go
+
+* Type and Method declarations, Struct Literals
+.play intro_concurrency/firststruct.go /Uniform/,$
+
+* Slices and range
+.play intro_concurrency/slices.go /init/,$
+
+* External packages
+.play intro_concurrency/extpack.go
+
+Download the external package:
+
+   go get github.com/gonum/optimize/functions
+
+* Putting them all together: Monte Carlo
+.play intro_concurrency/mc.go /func main/,$
+
+* Interfaces
+An *interface* is a set of methods
+
+  type Distribution interface {
+      Rand() float64
+  }
+
+A type *satisfies* the inteface if it has those methods
+
+  type Uniform struct{
+      Min float64
+      Max float64
+  }
+
+  func (u Uniform) Rand() float64 {
+      return rand.Float64()*(u.Max - u.Min) + u.Min
+  }
+
+Uniform satisfies the interface _implicitly_
+
+* Monte Carlo with interfaces
+.play intro_concurrency/interface.go /MonteCarlo estimates/,$
+
+* Concurrency
+
+* Concurrency
+
+MPI:
+
+- One source code, but multiple instances running
+- One thread per instance
```

Comment:
I believe the real answer to this question is that the set of things you can do with MPI is much larger than the set of things that's most frequently done with MPI. 

I can put a (frequently) at the end of the bullet.